crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1340,21 +1340,55 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Resolves the length value of a type.
+    ///
+    /// This is used to determine the value that would be returned
+    /// when `len(x)` is called on an object `x`.
+    fn len(&self, db: &'db dyn Db) -> Len {
+        match self {
+            Type::StringLiteral(string) => string.as_str(db).chars().count().into(),
+            Type::BytesLiteral(bytes) => bytes.len(db).into(),
+
+            Type::Tuple(tuple) => {
+                let elements = tuple.elements(db);
+
+                if elements.iter().any(|it| matches!(it, Type::Todo(..))) {
+                    Len::StaticallyUnknown
+                } else {
+                    elements.len().into()
+                }
+            }
+
```

Comment:
We should not have special handling for the `Todo` type as a tuple element, even if this gives us wrong results currently in some cases with starred elements. We can mark those cases with TODOs in the tests. They will be fixed when we fix the todo in tuple construction.

```suggestion
            Type::Tuple(tuple) => tuple.elements(db).len().into(),
```